From bd280abbaa489e599faf1183dc81af90095ac747 Mon Sep 17 00:00:00 2001
From: Peter Dinda <pdinda@northwestern.edu>
Date: Fri, 23 Jun 2017 06:33:17 -0500
Subject: [PATCH 14/20] Minor tweaks

- Makefile disables more gcc undefined behavior insanity
    - -fno-delete-null-pointer-checks applies throughout
    - -fno-isolate-erroneous-paths-attribute
      -fno-isolate-erroneous-paths-dereference are introduced
      in the comments - if your compiler supports them, add
      them.   Outside of this your compiler might add surprise
      "undefined" instructions on what it thinks is a null pointer
      dereference, thus killing you whenever you want to look
      at physical address zero.   It may also kill you at other
      times just for fun since, obviously, you can't be trusted
      by the language lawyers.
- percpu uses printk directly since ERROR_PRINT not available at this point
- corrected spelling in copyrights
---
 Makefile                  | 24 ++++++++++++++++++++++--
 include/nautilus/bug.h    |  2 +-
 include/nautilus/percpu.h |  4 ++--
 src/nautilus/chardev.c    |  2 +-
 4 files changed, 26 insertions(+), 6 deletions(-)

diff --git a/Makefile b/Makefile
index 2c5f5c4..12534af 100644
--- a/Makefile
+++ b/Makefile
@@ -308,15 +308,30 @@ NAUT_INCLUDE      := -D__NAUTILUS__ -Iinclude \
 
 CPPFLAGS        := $(NAUT_INCLUDE) -D__NAUTILUS__
 
-COMMON_FLAGS := -O2 \
+COMMON_FLAGS :=            -O2 \
 			   -fno-omit-frame-pointer \
 			   -ffreestanding \
 			   -fno-stack-protector \
 			   -fno-strict-aliasing \
                            -fno-strict-overflow \
+                           -fno-delete-null-pointer-checks \
 			   -mno-red-zone \
 			   -mcmodel=large
-			  
+
+#
+# Add these for more recent compilers to avoid having
+# the compiler insert surprise ud2 instructions for you should
+# you derefence address zero
+#
+#                           -fno-isolate-erroneous-paths-attribute \
+#                           -fno-isolate-erroneous-paths-dereference \
+#			  
+#
+# 
+# For testing, optionally add -fsanitize=undefined
+#
+#
+#
 
 CXXFLAGS := $(COMMON_FLAGS) \
 			-fno-exceptions \
@@ -342,6 +357,11 @@ CFLAGS:=   $(COMMON_FLAGS) \
 		   #-Wmissing-prototypes \
 		   #-Wstrict-prototypes \
 
+#
+#                   -Wextra \
+#                   -Wpedantic \
+#
+
 # NOTE: We MUST have max-page-size set to this here. Otherwise things
 # go off the rails for the Grub multiboot setup because the linker
 # does strange things...
diff --git a/include/nautilus/bug.h b/include/nautilus/bug.h
index a4fee2d..1ef74d9 100644
--- a/include/nautilus/bug.h
+++ b/include/nautilus/bug.h
@@ -8,7 +8,7 @@
  * led by Sandia National Laboratories that includes several national 
  * laboratories and universities. You can find out more at:
  * http://www.v3vee.org  and
- * http://xtack.sandia.gov/hobbes
+ * http://xstack.sandia.gov/hobbes
  *
  * Copyright (c) 2015, Kyle C. Hale <kh@u.northwestern.edu>
  * Copyright (c) 2015, The V3VEE Project  <http://www.v3vee.org> 
diff --git a/include/nautilus/percpu.h b/include/nautilus/percpu.h
index d2e582e..5e3c9fc 100644
--- a/include/nautilus/percpu.h
+++ b/include/nautilus/percpu.h
@@ -8,7 +8,7 @@
  * led by Sandia National Laboratories that includes several national 
  * laboratories and universities. You can find out more at:
  * http://www.v3vee.org  and
- * http://xtack.sandia.gov/hobbes
+ * http://xstack.sandia.gov/hobbes
  *
  * Copyright (c) 2015, Kyle C. Hale <kh@u.northwestern.edu>
  * Copyright (c) 2015, The V3VEE Project  <http://www.v3vee.org> 
@@ -114,7 +114,7 @@ do { \
             __per_cpu_put(var,newval,8); \
             break; \
         default: \
-            ERROR_PRINT("undefined op size in per_cpu_put\n"); \
+            printk("ERROR: undefined op size in per_cpu_put\n"); \
     } \
      } while (0)
 
diff --git a/src/nautilus/chardev.c b/src/nautilus/chardev.c
index ae93035..6369746 100644
--- a/src/nautilus/chardev.c
+++ b/src/nautilus/chardev.c
@@ -8,7 +8,7 @@
  * led by Sandia National Laboratories that includes several national 
  * laboratories and universities. You can find out more at:
  * http://www.v3vee.org  and
- * http://xtack.sandia.gov/hobbes
+ * http://xstack.sandia.gov/hobbes
  *
  * Copyright (c) 2016, Peter Dinda <pdinda@northwestern.edu>
  * Copyright (c) 2015, The V3VEE Project  <http://www.v3vee.org> 
-- 
1.9.1

